lib/types.js

```diff
@@ -296,22 +301,24 @@ module.exports = function (expect) {
             var type = this;
             keys.forEach(function (key, index) {
                 output.nl(index > 0 ? 1 : 0).i().block(function () {
+                    var valueActual = type.valueForKey(actual, key);
+                    var valueExpected = type.valueForKey(expected, key);
                     var valueOutput;
                     var annotation = output.clone();
-                    var conflicting = !equal(actual[key], expected[key]);
+                    var conflicting = !equal(valueActual, valueExpected);
                     var isInlineDiff = false;
                     if (conflicting) {
-                        if (!(key in expected)) {
+                        if (expectedKeys.indexOf(key) === -1) {
```

Comment:
This is potentially slow. We should consider making a temporary key index or add type methods to ask if a key exist: `hasKey`.